library/alloc/src/boxed.rs

```diff
@@ -472,13 +720,30 @@ impl<T: ?Sized> Box<T> {
     )]
     #[inline]
     #[doc(hidden)]
-    pub fn into_unique(b: Box<T>) -> Unique<T> {
+    pub fn into_unique(b: Self) -> (Unique<T>, A) {
         // Box is recognized as a "unique pointer" by Stacked Borrows, but internally it is a
         // raw pointer for the type system. Turning it directly into a raw pointer would not be
         // recognized as "releasing" the unique pointer to permit aliased raw accesses,
-        // so all raw pointer methods go through `leak` which creates a (unique)
-        // mutable reference. Turning *that* to a raw pointer behaves correctly.
-        Box::leak(b).into()
+        // so all raw pointer methods have to leak the box. Turning *that* to a raw pointer
+        // behaves correctly.
+        let b = mem::ManuallyDrop::new(b);
+
+        // The box is unitiliazed later when moving out the allocator. The pointer is stored
+        // beforehand.
+        let ptr = b.0;
+        let alloc = unsafe { ptr::read(&b.1) };
+        (ptr, alloc)
+    }
+
+    /// Returns a reference to the underlying allocator.
+    ///
+    /// Note: this is an associated function, which means that you have
+    /// to call it as `Box::alloc_ref(&b)` instead of `b.alloc_ref()`. This
+    /// is so that there is no conflict with a method on the inner type.
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[inline]
+    pub fn alloc_ref(b: &Self) -> &A {
```

Comment:
To me, this sounds like we're allocating a reference, not obtaining a reference to the allocator.

Reply:
`allocator` was also suggested in https://github.com/rust-lang/rust/pull/77187#discussion_r509881386, which might be clearest.

Reply:
I'm fine with both, I just picked `alloc_ref` because of the trait name. Maybe we should discuss (in the WG), if we should rename `AllocRef` to `Allocator` and `alloc` to `allocate`.

Reply:
Proposal opened: https://github.com/rust-lang/wg-allocators/issues/76